File: platformio/test/com/jetbrains/cidr/cpp/embedded/platformio/project/PlatformioBoardsParserTest.java
package com.jetbrains.cidr.cpp.embedded.platformio.project;

import com.intellij.testFramework.UsefulTestCase;
import com.intellij.util.ResourceUtil;

import javax.swing.tree.TreeNode;
import java.util.Enumeration;
import java.util.stream.Stream;

import static com.jetbrains.cidr.cpp.embedded.platformio.project.BoardInfo.EMPTY;
import static com.jetbrains.cidr.cpp.embedded.platformio.project.DeviceTreeNode.TYPE.BOARD;
import static com.jetbrains.cidr.cpp.embedded.platformio.project.DeviceTreeNode.TYPE.FRAMEWORK;
import static com.jetbrains.cidr.cpp.embedded.platformio.project.DeviceTreeNode.TYPE.VENDOR;
import static com.jetbrains.cidr.cpp.embedded.platformio.project.SourceTemplate.ARDUINO;
import static com.jetbrains.cidr.cpp.embedded.platformio.project.SourceTemplate.GENERIC;
import static java.util.Collections.list;

public class PlatformioBoardsParserTest extends UsefulTestCase {

  private String myJson;

  @Override
  protected void setUp() throws Exception {
    super.setUp();
    myJson = ResourceUtil.loadText(PlatformioBoardsParserTest.class.getResourceAsStream(this.getClass().getSimpleName() + ".json"));
  }

  public void testParse() {
    DeviceTreeNode root = BoardsJsonParser.parse(myJson);
    assertEquals(190, root.getChildCount());

    DeviceTreeNode board = childrenStream(root)
      .filter(node -> node.hasSameValues("ST", VENDOR, EMPTY))
      .flatMap(node -> childrenStream(node))
      .filter(node -> node
        .hasSameValues("ST 32F3348DISCOVERY", BOARD, new BoardInfo(GENERIC, "--board", "disco_f334c8")))
      .findAny()
      .orElseThrow(AssertionError::new);
    assertEquals(2, board.getChildCount());

    assertTrue(board.getChildAt(0).hasSameValues("mbed", FRAMEWORK,
                                                 new BoardInfo(GENERIC, "--board", "disco_f334c8", "-O",
                                                               "framework=mbed")));
    assertTrue(board.getChildAt(1).hasSameValues("stm32cube", FRAMEWORK,
                                                 new BoardInfo(GENERIC, "--board", "disco_f334c8", "-O",
                                                               "framework=stm32cube")));
    DeviceTreeNode arduinoBoard = childrenStream(root)
      .filter(node -> node.hasSameValues("Armed", VENDOR, EMPTY))
      .flatMap(node -> childrenStream(node))
      .filter(node -> node
        .hasSameValues("3D Printer Controller", BOARD, new BoardInfo(ARDUINO, "--board", "armed_v1")))
      .findAny()
      .orElseThrow(AssertionError::new);
    assertTrue(arduinoBoard.getChildAt(0).hasSameValues("arduino", FRAMEWORK,
                                                        new BoardInfo(ARDUINO, "--board", "armed_v1", "-O",
                                                                      "framework=arduino")));
  }

  @SuppressWarnings({"unchecked", "rawtypes"})
  private static Stream<DeviceTreeNode> childrenStream(final TreeNode treeNode) {
    return list((Enumeration)treeNode.children()).stream();
  }
}

File: platformio/test/com/jetbrains/cidr/cpp/embedded/platformio/project/PlatformioBoardsParserTest.kt
package com.jetbrains.cidr.cpp.embedded.platformio.project

import com.intellij.testFramework.UsefulTestCase
import com.intellij.util.ResourceUtil
import com.jetbrains.cidr.cpp.embedded.platformio.project.DeviceTreeNode.TYPE.*

class PlatformioBoardsParserTest : UsefulTestCase() {
  private var myJson: String? = null

  @Throws(Exception::class)
  override fun setUp() {
    super.setUp()
    myJson = ResourceUtil.loadText(
      this.javaClass.getResourceAsStream(this.javaClass.simpleName + ".json")!!)
  }

  private fun boardInfo(sourceTemplate: SourceTemplate, vararg params: String) = BoardInfo(sourceTemplate, listOf(*params))
  fun testParse() {
    val root = BoardsJsonParser.parse(myJson!!)
    assertEquals(190, root.childCount)
    val board = root.children().asSequence()
      .filter { node: DeviceTreeNode -> node.hasSameValues("ST", VENDOR, SourceTemplate.EMPTY_BOARD_INFO) }
      .flatMap { node: DeviceTreeNode -> node.children().asSequence() }
      .single { node: DeviceTreeNode -> node.name.startsWith("ST 32F3348DISCOVERY") }
    assertTrue(
      board.hasSameValues("ST 32F3348DISCOVERY (STM32F334C8T6, 72MHz, ROM: 64K, RAM: 12K)", BOARD,
                          boardInfo(SourceTemplate.GENERIC, "--board", "disco_f334c8")))
    assertEquals(2, board.childCount)
    assertTrue(board.getChildAt(0).hasSameValues("mbed", FRAMEWORK,
                                                 boardInfo(SourceTemplate.GENERIC, "--board", "disco_f334c8", "-O",
                                                           "framework=mbed")))
    assertTrue(board.getChildAt(1).hasSameValues("stm32cube", FRAMEWORK,
                                                 boardInfo(SourceTemplate.GENERIC, "--board", "disco_f334c8", "-O",
                                                           "framework=stm32cube")))
    val arduinoBoard = root.children().asSequence()
      .filter { node: DeviceTreeNode -> node.hasSameValues("Armed", VENDOR, SourceTemplate.EMPTY_BOARD_INFO) }
      .flatMap { node: DeviceTreeNode -> node.children().asSequence() }
      .single { node: DeviceTreeNode -> node.name.startsWith("3D Printer Controller") }
    arduinoBoard.hasSameValues("3D Printer Controller (STM32F407VET6, 168MHz, ROM: 512K, RAM: 192K)", BOARD,
                               boardInfo(SourceTemplate.ARDUINO, "--board", "armed_v1"))
    assertTrue(arduinoBoard.getChildAt(0).hasSameValues("arduino", FRAMEWORK,
                                                        boardInfo(SourceTemplate.ARDUINO, "--board", "armed_v1", "-O",
                                                                  "framework=arduino")))
  }

}